controlpanel/frontend/views/app.py

```diff
@@ -232,16 +245,28 @@ def post(self, request, *args, **kwargs):
         return super().post(request, *args, **kwargs)
 
 
-class AddCustomers(UpdateApp):
+class AddCustomers(LoginRequiredMixin, PermissionRequiredMixin, UpdateView):
+    form_class = AddAppCustomersForm
+    model = App
     permission_required = 'api.add_app_customer'
 
-    def perform_update(self, **kwargs):
-        app = self.get_object()
-        emails = self.request.POST.get('customer_email')
-        emails = re.split(r'[,; ]+', emails)
-        emails = [email.strip() for email in emails]
-        app.add_customers(emails)
+    def form_invalid(self, form):
+        self.request.session['add_customer_form_errors'] = form.errors
+        return HttpResponseRedirect(
+            reverse_lazy("manage-app", kwargs={"pk": self.kwargs['pk']}),
+        )
+
+    def form_valid(self, form):
+        self.get_object().add_customers(form.cleaned_data['customer_email'])
+        return HttpResponseRedirect(self.get_success_url())
+
+    def get_form_kwargs(self):
+        kwargs = FormMixin.get_form_kwargs(self)
```

Comment:
are you doing FormMixin.get_form_kwargs because `self.get_form_kwargs` comes from ModelFormMixin?

Reply:
Yes, because AddAppCustomersForm is not a subclass of ModelForm and making it one would be more complex.